FutureNorthants/northants-website
# Added RSS feed generation to frontend

### Setup
You will need the relevant backend branch installed: https://github.com/FutureNorthants/northants-website/pull/182
Now stop your frontend server, pull this branch into your frontend and run `yarn codegen` and it should create some new generated files. If it doesn't, let me know. Run `yarn dev` to restart the server.

### Testing
1. With another terminal window open in your frontend, go to `localhost:3000/rss/feed.xml` and then take a look in the `public` folder of your frontend. You should find it has created an `rss/__generated__` folder and that there are 4 files in there - `feed.xml`, `atom.xml`, `feed.json` and `updated.json`, and the browser has returned the feed.xml file to you.
2. Verify that `updated.json` contains a bit of json like this: `{"date":"something"}` and that the something is the date a few seconds ago.
3. Verify that the other three files include sane looking data in the correct format for the three types of content feed supported ([RSS2](https://validator.w3.org/feed/docs/rss2.html#requiredChannelElements), [Atom](https://datatracker.ietf.org/doc/html/rfc4287) and [JSON Feed](https://www.jsonfeed.org/version/1.1/)).
4. View the other two types of feed file in your browser and verify they look OK there - `localhost:3000/rss/atom.xml` and `localhost:3000/rss/feed.json`.
5. Verify that the 4 files have not changed date during all this (haven't been regenerated).
6. Delete the `updated.json` file or edit it to have a date several weeks or months in the past. Refresh the browser page. Verify that the 4 files have a current date (have been regenerated).
7. Create a new news article in the CMS and publish it.
8. Refresh the frontend browser page. Verify the 4 files in `rss/__generated__` have been regenerated and now include the new news story.
9. Install an RSS reader, e.g. An Otter RSS, and try adding a new feed, pointing it at the address of your local frontend, e.g. `http://localhost:3000`, it should pick up the three possible feed files to use via the autodiscover headers. Subscribe to the feed and check you can see news stories via the reader.
10. Try and view an unsupported filename in your browser e.g. `localhost:3000/rss/imadeitup.xml` and verify you get an appropriate error page.
